androguard/core/axml/__init__.py

```diff
@@ -1527,22 +1527,50 @@ def __init__(self, raw_buff:bytes) -> None:
                     elif pkg_chunk_header.type == RES_TABLE_TYPE_TYPE:
                         # Parse a RES_TABLE_TYPE
                         # http://androidxref.com/9.0.0_r3/xref/frameworks/base/tools/aapt2/format/binary/BinaryResourceParser.cpp#311
+                        start_of_chunk = self.buff.tell() - 8
+                        expected_end_of_chunk = start_of_chunk + pkg_chunk_header.size
                         a_res_type = ARSCResType(self.buff, pc)
                         self.packages[package_name].append(a_res_type)
                         self.resource_configs[package_name][a_res_type].add(a_res_type.config)
 
                         logger.debug("Config: {}".format(a_res_type.config))
 
                         entries = []
+                        FLAG_OFFSET16 = 0x02
+                        NO_ENTRY_16 = 0xFFFF
+                        NO_ENTRY_32 = 0xFFFFFFFF
+                        expected_entries_start = start_of_chunk + a_res_type.entriesStart
+
+                        # Helper function to convert 16-bit offset to 32-bit
+                        def offset_from16(off16):
+                            return NO_ENTRY_16 if off16 == NO_ENTRY_16 else off16 * 4
+
                         for i in range(0, a_res_type.entryCount):
                             current_package.mResId = current_package.mResId & 0xffff0000 | i
-                            entries.append((unpack('<i', self.buff.read(4))[0], current_package.mResId))
+                            # Check if FLAG_OFFSET16 is set
+                            if a_res_type.flags & FLAG_OFFSET16:
+                                # Read as 16-bit offset
+                                offset_16 = unpack('<H', self.buff.read(2))[0]
+                                offset = offset_from16(offset_16)
+                                if offset == NO_ENTRY_16:
+                                    continue
+                            else:
+                                # Read as 32-bit offset
+                                offset = unpack('<I', self.buff.read(4))[0]
+                                if offset == NO_ENTRY_32:
+                                    continue
+                            entries.append((offset, current_package.mResId))
 
                         self.packages[package_name].append(entries)
-
-                        for entry, res_id in entries:
-                            if entry != -1:
-                                ate = ARSCResTableEntry(self.buff, res_id, pc)
+                        
+                        base_offset = self.buff.tell()
+                        if base_offset != expected_entries_start:
+                            # FIXME: seems like I am missing 2 bytes here in some cases, though it does not affect the result
```

Comment:
This occasional 2-byte difference is expected as entries are aligned to 4-byte boundaries. It should be easy to account for that instead of producing a warning.

Reply:
Yeah, IIRC your code from before should already have handled the APKs correctly. I meant you could emit the warning only if the expected start doesn't match the *position adjusted to account for alignment* here. I did mention the alignment before (and linked to my own code implementing `FLAG_OFFSET16,` `FLAG_COMPACT`, and `FLAG_SPARSE`) in my comments in 1066 :)

Reply:
I am checking this further and the misalignment seems to happen right before starting to parse the actual entries and right after having parsed the offsets to which the entries are located. As far as [I could tell](https://android.googlesource.com/platform/frameworks/base/+/refs/heads/android14-release/libs/androidfw/include/androidfw/ResourceTypes.h#1456) all cases parsing the offsets are 4-byte-aligned already, so I am missing/misunderstanding something here.

Reply:
The offsets are always 4-byte-aligned. And the header is followed by an array of offsets, followed by the entries themselves.

Before the introduction of 16-bit offsets the size of both the header and array would always be a multiple of 32 bits, meaning the end of the array would always be 4-byte-aligned and there would not be a gap between the end of the array and the 4-byte-aligned start of the entries.

But with 16-bit offsets there can be a 2-byte/16-bit gap between the (now 2-byte-aligned) end of the array and the 4-byte-aligned start of the entries because the array size is no longer always a multiple of 32 bits.

Reply:
great! thank you for explaining!